kafka/consumer.go

```diff
@@ -0,0 +1,393 @@
+				}
+			}
+
+			_, newest, err := c.consumer.QueryWatermarkOffsets(c.conf.Topics[0], partition, c.conf.MetadataTimeout)
+			if err != nil {
+				log.Error(3, "kafka-consumer: Error when querying for offsets: %s", err)
+			} else {
+				c.partitionLogSize[partition].Set(int(newest))
+			}
+
+			c.partitionOffset[partition].Set(int(offset))
+			if err == nil {
+				lag := int(newest - offset)
+				c.partitionLag[partition].Set(lag)
+				if c.LagMonitor != nil {
+					c.LagMonitor.StoreLag(partition, lag)
+				}
+			}
+		}
+	}
+
+	ticker := time.NewTicker(c.conf.OffsetCommitInterval)
+	for {
+		select {
+		case ts := <-ticker.C:
+			storeOffsets(ts)
+		case <-c.stopChan:
+			storeOffsets(time.Now())
+			return
+		}
+	}
+}
+
+func (c *Consumer) startConsumer() error {
+	var offset confluent.Offset
+	var err error
+	var topicPartitions confluent.TopicPartitions
+	c.currentOffsets = make(map[int32]*int64, len(c.Partitions))
+
+	for i, topic := range c.conf.Topics {
+		for _, partition := range c.Partitions {
+			var currentOffset int64
+			switch c.conf.StartAtOffset {
+			case "oldest":
+				currentOffset, _, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetBeginning), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			case "newest":
+				_, currentOffset, err = c.tryGetOffset(topic, partition, int64(confluent.OffsetEnd), 3, time.Second)
+				if err != nil {
+					return err
+				}
+			default:
+				offsetDuration, err := time.ParseDuration(c.conf.StartAtOffset)
+				if err != nil {
+					return fmt.Errorf("invalid offest format %s: %s", c.conf.StartAtOffset, err)
+				}
+				currentOffset = time.Now().Add(-1*offsetDuration).UnixNano() / int64(time.Millisecond)
+				currentOffset, _, err = c.tryGetOffset(topic, partition, currentOffset, 3, time.Second)
+				if err != nil {
+					return err
+				}
+			}
+
+			offset, err = confluent.NewOffset(currentOffset)
+			if err != nil {
+				return err
+			}
+
+			topicPartitions = append(topicPartitions, confluent.TopicPartition{
+				Topic:     &topic,
+				Partition: partition,
+				Offset:    offset,
+			})
+
+			if i == 0 {
+				c.currentOffsets[partition] = &currentOffset
+			}
+		}
+	}
+
+	return c.consumer.Assign(topicPartitions)
+}
+
+func (c *Consumer) tryGetOffset(topic string, partition int32, offsetI int64, attempts int, sleep time.Duration) (int64, int64, error) {
+	offset, err := confluent.NewOffset(offsetI)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	var val1, val2 int64
+
+	attempt := 1
+	for {
+		if offset == confluent.OffsetBeginning || offset == confluent.OffsetEnd {
+			val1, val2, err = c.consumer.QueryWatermarkOffsets(topic, partition, c.conf.MetadataTimeout)
+		} else {
+			times := []confluent.TopicPartition{{Topic: &topic, Partition: partition, Offset: offset}}
+			times, err = c.consumer.OffsetsForTimes(times, c.conf.MetadataTimeout)
+			if err == nil {
+				if len(times) == 0 {
+					err = fmt.Errorf("Got 0 topics returned from broker")
```

Comment:
we need to fall back to offsetEnd

Reply:
why should this fall back to offsetEnd, while in the above case it should fall back to offsetBeginning? shouldn't they both fall back to offsetBeginning?

Reply:
how is that? https://github.com/grafana/metrictank/blob/985f69a32671b8769ddcababaf5d7be7059b21c8/kafka/consumer.go#L360-L383

Reply:
yep, i meant offsetBeginning.  LGTM